### ISSUE TITLE
VIPS commandline options for file conversions

### ISSUE THREAD

**Issue by sachinwalia2k8** [jcupitt/libvips#515]
As per this link http://www.vips.ecs.soton.ac.uk/supported/7.42/doc/html/libvips/using-cli.html
file conversion is supported using various foreign libraries. For example to output jpg this seems to work:

`vips affine k2.jpg x.jpg[Q=90,strip] "2 0 0 1"`

How do I check what all flags/options are supported by different foreign libraries.

Also is there some comprehensive list of image formats that are supported?

thanks.

**Comment by jcupitt** [comment]
Yes, check the docs here:

http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsForeignSave.html

You can look them up with the `vips` program too, for example:

```
$ vips jpegsave
save image to jpeg file
usage:
   jpegsave in filename
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   Q            - Q factor, input gint
			default: 75
			min: 1, max: 100
   profile      - ICC profile to embed, input gchararray
   optimize-coding - Compute optimal Huffman coding tables, input gboolean
			default: false
   interlace    - Generate an interlaced (progressive) jpeg, input gboolean
			default: false
   no-subsample - Disable chroma subsample, input gboolean
			default: false
   trellis-quant - Apply trellis quantisation to each 8x8 block, input gboolean
			default: false
   overshoot-deringing - Apply overshooting to samples with extreme values, input gboolean
			default: false
   optimize-scans - Split the spectrum of DCT coefficients into separate scans, input gboolean
			default: false
   strip        - Strip all metadata from image, input gboolean
			default: false
   background   - Background value, input VipsArrayDouble
```

**Comment by sachinwalia2k8** [comment]
Thanks @jcupitt. Few more questions:

1. I see that there is a corresponding load function such as jpegload to read image. What is the purpose of this function. Is it applicable to situation when you are just doing a file conversion from let's say png to jpeg?

2. How does it format conversion internally works. Does it relies on ImageMagick / GraphicsMagick? If not then are there any tradeoffs in using VIPS vs any of these tools as VIPS is very fast compared to IM/GM.

3. How do we support newer format such as FLIF through VIPS. (https://github.com/FLIF-hub/FLIF)

**Comment by jcupitt** [comment]
Hi again, 

Here's how `new_from_file` works:

* break off any `[options]` from the filename
* find all subclasses of `VipsForeignLoad` and sort by `@priority`
* offer the filename to the `is_a` class method of every class in turn, stopping when we get a hit
* call the `new` method of that class, passing in the filename, plus any options

So the command:

```
$ vips invert x.jpg y.tif
```

Creates a pipeline of `jpegload` -> `invert` -> `tiffsave`.

If you want to add a new format to vips, you need to define a new subclass of `VipsForeignLoad`. There are some notes in the docs:

http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/VipsForeignSave.html

Scroll down to "description".

There's an operation called `magickload` which will load an image via ImageMagick. This is a low priority loader, so it acts as a fallback if everything else fails. Formats like BMP, with no native vips loader, will be loaded by this. 

As you say, it is rather slow --- the native vips loaders are generally a lot quicker. You can benchmark them like this:

```
$ time vips magickload wtc.jpg x.ppm
real	0m1.826s
user	0m1.356s
sys	0m0.484s
$ time vips jpegload wtc.jpg x.ppm
real	0m1.315s
user	0m0.764s
sys	0m0.608s
```

So the vips one is about 2x faster (roughly).

**Comment by jcupitt** [comment]
There's an issue for FLIP support, perhaps you saw: https://github.com/jcupitt/libvips/issues/339

Once 8.4 is done, it'll probably get looked at.

**Comment by jcupitt** [comment]
I think we're done here, I'll close.
